src/validation/mod.rs

```diff
@@ -162,94 +164,96 @@ pub fn validate(wasm: &[u8]) -> Result<ValidationInfo> {
-    //             }
-    //             4 => {
-    //                 // type funcref
-    //             }
-    //             5 => {
-    //                 // type reftype
-    //             }
-    //             6 => {
-    //                 // type reftype
-    //             }
-    //             7 => {
-    //                 // type reftype
-    //             }
-    //             _ => unimplemented!()
-    //         }
-    //      }
-    //     todo!("element section not yet supported")
-    // })?;
-    // #endregion
-
-    let _element: Option<()> =
-        handle_section(&mut wasm, &mut header, SectionTy::Element, |_, _| {
-            todo!("element section not yet supported")
-        })?;
-    while (skip_section(&mut wasm, &mut header)?).is_some() {}
+    let elements: Vec<ElemType> = handle_section(&mut wasm, &mut header, SectionTy::Element, |wasm, _| {
+        use crate::core::reader::types::element::*;
+        use crate::RefType;
+        let mut elem_vec: Vec<ElemType> = Vec::new();
+        // https://webassembly.github.io/spec/core/binary/modules.html#element-section
+        // TODO: replace with wasm.read_vec in the future
+        let vec_length = wasm.read_var_u32().unwrap();
+        trace!("Element sections no.: {}", vec_length);
+        for i in 0..vec_length {
+            let ty = wasm.read_var_u32().unwrap();
+            // https://webassembly.github.io/spec/core/syntax/modules.html#element-segments
+            // https://webassembly.github.io/spec/core/binary/modules.html#element-section
+            // We can treat the ttype as a 3bit integer
+            // If it's not 3 bits I am not sure what to do
+            // bit 0 => diff between passive|declartive and active segment
+            // bit 1 => presence of an explicit table index for an active segment
+            // bit 2 => use of element type and element expressions instead of element kind and element indices
+            assert!((ty & 0b111) <= 0b111, "Element section is not encoded correctly. The type of this element is over 7 (0b111)");
+            // decide if we should
+            let elem_mode = if ty & 0b001 == 0b001 {
+                if ty & 0b010 == 0b010 {
+                    ElemMode::Declarative
+                } else {
+                    ElemMode::Passive
+                }
+            } else {
+                let table_idx = if ty & 0b010 == 0b010 {
+                    wasm.read_var_u32()?
+                } else {
+                    0
+                };
+                if tables.len() <= table_idx as usize {
+                    return Err(Error::UnknownTable);
+                }
+                let expr = read_constant_expression::read_constant_instructions(wasm, None, None)?;
+
+                ElemMode::Active(ActiveElem {
+                    table: table_idx,
+                    offset: expr,
+                })
+            };
+            let use_of_el_ty_and_el_exprs = ty & 0b100 == 0b100;
```

Comment:
The name is really verbose yet I fail to understand what it is doing xD
This isn't an issue with the code per se, more of an issue for my understanding of bit 3 (see previous comment)